presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcClient.java

```diff
@@ -717,6 +722,10 @@ public WriteMapping toWriteMapping(ConnectorSession session, Type type)
             return WriteMapping.sliceMapping(dataType, longDecimalWriteFunction(decimalType));
         }
 
+        if (type == TIMESTAMP) {
+            return WriteMapping.longMapping("timestamp", timestampWriteFunction(session));
```

Comment:
`timestampWriteFunction` is correct (uses `j.t.LocalDateTime` so it doesn't lose information), but it's not compatible with all the JDBC drivers. For example, it will work with PostgreSQL driver, but will not work with SQL Server driver.

I think the right approach is to add this mapping on case-by-case basis. I.e. not in this class, but in specific classes instead. For example, for PostgreSQL this would be in `io.prestosql.plugin.postgresql.PostgreSqlClient#toWriteMapping` (it's already there btw). Tests should go in `Test*TypeMapping` class.

Reply:
Adding the mapping on a case-by-case basis sounds sensible, but I have a slight concern that this could lead to a lot of code duplication if more JDBC-based connectors were to be added in the future. In any case I'll update the PR.

Just as a note too it appears that we don't have any type mapping tests for SQL Server at all.

Also, looks like `LocalDateTime` may be supported by `mssql-jdbc` since [`v7.1.0`](https://github.com/microsoft/mssql-jdbc/releases/tag/v7.1.0).

Reply:
> Just as a note too it appears that we don't have any type mapping tests for SQL Server at all.

I'm aware.
Initially, we didn't have tools to run this type of tests on SQL Server. Now we have everything we need, so you can add this if you wish. `TestPostgreSqlTypeMapping` is the best starting point.

> Adding the mapping on a case-by-case basis sounds sensible, but I have a slight concern that this could lead to a lot of code duplication if more JDBC-based connectors were to be added in the future

I understand. But the mapping actually varies from connector to connector

- some drivers accept `java.sql.Timestamp` only
- some accept `java.time.LocalDateTime`
- some accept `java.time.LocalDateTime` but something is broken and you need to use something else (see `io.prestosql.plugin.postgresql.PostgreSqlClient#timestampWriteFunction`)

Also, mapping is usually simple. There is a lot more test code for this. We consciously have code duplication for tests too, as there are slight differences.